docs/4.x/tests-in-depth-faq.md

```diff
@@ -0,0 +1,144 @@
+---
+category: DevelopInDepth
+---
+# Tests FAQ 
+
+## How do I modify the config ini to test certain conditions?
+
+### Integration and system PHP tests
+
+You can use the [configuration object](/guides/piwiks-ini-configuration) as usual and set any specific ini setting before executing the test.
+
+You don't need to reset the ini after each test as this is done automatically for you.
+
+You can find an example in the [BlobReportLimitingTest](https://github.com/matomo-org/matomo/blob/4.4.0/tests/PHPUnit/System/BlobReportLimitingTest.php#L194-L216). 
+
+### UI tests
+
+You can overwrite the config like below:
+
+```
+testEnvironment.configOverride.General = {
+    browser_archiving_disabled_enforce: '1',
+    enable_browser_archiving_triggering: '0',
+};
+testEnvironment.save();
+```
+ 
+## How do I change objects or configurations in tests?
+
+### Generic dependency injection overwrite for tests
+
+You can use either `/config/environment/test.php` or a plugin specific `config/test.php` say `plugins/CoreHome/config/test.php` to overwrite any dependency in Matomo during tests.
+
+For examples see [config/test.php](https://github.com/matomo-org/matomo/blob/4.4.0/config/environment/test.php) and [Tour/test.php](https://github.com/matomo-org/matomo/blob/4.4.0/plugins/Tour/config/test.php).
+
+Changes here are usually applied to all tests.
+
+### Integration and system PHP tests
+
+If you want to apply changes only for a specific test then you can overwrite the `provideContainerConfigBeforeClass` or the `provideContainerConfig` method to change the injected dependencies if neeeded.
```

Comment:
not 100% sure when to use which of the two methods. When would we use one or the other?

Reply:
provideContainerConfig for IntegrationTestCase and UnitTestCase classes, provideContainerConfigBeforeClass for SystemTestCase, since system tests don't do the reset logic as mentioned above. The override will be for an entire test case though, not just a single test.

Reply:
great, thanks 👍